tests/cupyx_tests/scipy_tests/sparse_tests/test_linalg.py

```diff
@@ -395,3 +398,174 @@ def test_invalid(self):
+
+        class BaseMatlike(interface.LinearOperator):
+            args = ()
+
+            def __init__(self, dtype):
+                self.dtype = cp.dtype(dtype)
+                self.shape = original.shape
+
+            def _matvec(self, x):
+                return mv(x, self.dtype)
+
+        class HasRmatvec(BaseMatlike):
+            args = ()
+
+            def _rmatvec(self, x):
+                return rmv(x, self.dtype)
+
+        class HasAdjoint(BaseMatlike):
+            args = ()
+
+            def _adjoint(self):
+                shape = self.shape[1], self.shape[0]
+                matvec = partial(rmv, dtype=self.dtype)
+                rmatvec = partial(mv, dtype=self.dtype)
+                return interface.LinearOperator(matvec=matvec,
+                                                rmatvec=rmatvec,
+                                                dtype=self.dtype,
+                                                shape=shape)
+
+        class HasRmatmat(HasRmatvec):
+            def _matmat(self, x):
+                return original.dot(x)
+
+            def _rmatmat(self, x):
+                return original.T.conj().dot(x)
+
+        cases.append((HasRmatvec(dtype), original))
+        cases.append((HasAdjoint(dtype), original))
+        cases.append((HasRmatmat(dtype), original))
+        return cases
+
+    def _make_cases(self):
+        self.cases = []
+        original = cp.array([[1., 2., 3.], [4., 5., 6.]])
+        self.cases += self._define_cases(original, cp.float32)
+        self.cases += self._define_cases(original, cp.float64)
+        self.cases += [(interface.aslinearoperator(M).T, A.T)
+                       for M, A in self._define_cases(original.T, cp.float64)]
+        self.cases += [(interface.aslinearoperator(M).H, A.T.conj())
+                       for M, A in self._define_cases(original.T, cp.float64)]
+
+        original = cp.array([[1, 2j, 3j], [4j, 5j, 6]])
+        self.cases += self._define_cases(original, cp.complex_)
+        self.cases += [(interface.aslinearoperator(M).T, A.T)
+                       for M, A in self._define_cases(original.T, cp.complex_)]
+        self.cases += [(interface.aslinearoperator(M).H, A.T.conj())
+                       for M, A in self._define_cases(original.T, cp.complex_)]
+
+    def test_basic(self):
+        for M, A_array in self.cases:
+            A = interface.aslinearoperator(M)
+            M, N = A.shape
+
+            xs = [cp.array([1, 2, 3]),
+                  cp.array([[1], [2], [3]])]
+            ys = [cp.array([1, 2]), cp.array([[1], [2]])]
+
+            if A.dtype == cp.complex_:
+                xs += [cp.array([1, 2j, 3j]),
+                       cp.array([[1], [2j], [3j]])]
+                ys += [cp.array([1, 2j]), cp.array([[1], [2j]])]
+
+            x2 = cp.array([[1, 4], [2, 5], [3, 6]])
+
+            for x in xs:
+                cp.testing.assert_array_equal(A.matvec(x), A_array.dot(x))
```

Comment:
`testing.assert_array_equal` does not check dtype's equality. Could you use `testing.numpy_cupy_array_equal` decorator instead?

Reply:
hey @asi1024 , the comparison here is between two cupy arrays (of the same types to start with), the numpy-cupy comparison is done in the `test_sparse` and `test_dense` functions...